File: make-tech-article-data.js
import fetch from "node-fetch";
import dotenv from "dotenv";
import fs from "fs";
import xml2js from "xml2js";

const BEGIN_YEAR = 2023;
const END_YEAR = 2024;
const BEGIN_MONTH = 4;
const FILE_PATH = "./src/data/TechArticleData.ts";

//TODO: 取得するのは2年分
const main = async () => {
    console.log("#### Begin make article data ####");

    console.log("#### Lading Config... ####");
    const env = dotenv.config();
    console.log(env);
    console.log(env.parsed.QIITA_API_KEY);

    try {
        console.log("#### Fetch Zenn Info####");
        const originalZennData = await getZennArticles();
        const zennData = getZennData(originalZennData.articles);
        const zennArticleList = getZennArticleList(originalZennData.articles);
        console.log(zennArticleList.length);

        console.log("#### Fetch Qiita Info####");
        const originalQiitaData = await getQiitaArticles(env);
        const qiitaData = getQiitaData(originalQiitaData);
        const qiitaArticleList = getQiitaArticleList(originalQiitaData);
        console.log(qiitaArticleList.length);

        console.log("#### Fetch note Info####");
        const noteData = await getNoteArticles();
        console.log(noteData.length);

        console.log("#### Fetch SpeakerDeck Info####");
        const decks = await getSpeakerDecks();
        console.log(decks.length);

        console.log("#### Fetch GitHub Info####");
        const gitHubData = await getGitHubControbutions();
        const gitHubContributions = gitHubData.contributions.map(
            (contribution) => {
                return {
                    date: contribution.date.replace(/-/g, "/"),
                    contributionCount: contribution.contributionCount,
                };
            }
        );
        console.log(gitHubContributions.length);

        // データをソートする
        console.log("#### Make Sort And Paging Articles List ####");
        const maegedAndSortedArticleList = [
            ...zennArticleList,
            ...qiitaArticleList,
        ].sort(comparePublishDate);
        console.log(`Length is ${maegedAndSortedArticleList.length}`);
        const pagingArticles = getPagingArticles(maegedAndSortedArticleList);
        console.log(`Page Size is ${pagingArticles.length}`);
        const nowYear = new Date().getFullYear();
        const popularArticles = zennArticleList
            .filter((article) => article.year.includes(nowYear))
            .sort(function (a, b) {
                return b.likeCount - a.likeCount;
            });
        const rankEmojiArray = ["🥇", "🥈", "🥉"];
        const slicedPopularArticles =
            popularArticles.length > 2
                ? popularArticles.slice(0, 3)
                : popularArticles;
        const popularArticlesWithRank = slicedPopularArticles.map(
            (article, index) => {
                return {
                    ...article,
                    treeType: rankEmojiArray[index],
                };
            }
        );

        console.log("#### Make Chart Data ####");
        const date = new Date();
        const currentYear = date.getFullYear();
        const currentMonth = date.getMonth() + 1;
        const chartData = {
            articlesCounts: [],
            yearArticleCounts: [],
            favoritesCounts: [],
            yearFavoritesCounts: [],
        };
        for (let year = BEGIN_YEAR; year <= END_YEAR; year++) {
            const yearArticleCount = {
                year: year.toString(),
                articles: 0,
            };
            const yearFavoritesCount = {
                year: year.toString(),
                favorites: 0,
            };
            for (let month = 0; month < 12; month++) {
                //! 記事投稿を始めた時期より前はスキップ
                if (year === BEGIN_YEAR && month + 1 < BEGIN_MONTH) {
                    continue;
                }
                //! 未来になったら抜ける
                if (currentYear === year && currentMonth < month + 1) {
                    break;
                }
                const key = getFormattedDate(year, month + 1);
                const articlesCount = {
                    yearMonth: key,
                    zenn: zennData[key] ? zennData[key].articlesCount : 0,
                    qiita: qiitaData[key] ? qiitaData[key].articlesCount : 0,
                };
                chartData.articlesCounts.push(articlesCount);
                yearArticleCount.articles +=
                    articlesCount.zenn + articlesCount.qiita;

                const favoritesCount = {
                    yearMonth: key,
                    zenn: zennData[key] ? zennData[key].likeCount : 0,
                    qiita: qiitaData[key] ? qiitaData[key].likeCount : 0,
                };
                chartData.favoritesCounts.push(favoritesCount);
                yearFavoritesCount.favorites +=
                    favoritesCount.zenn + favoritesCount.qiita;
            }
            chartData.yearArticleCounts.push(yearArticleCount);
            chartData.yearFavoritesCounts.push(yearFavoritesCount);
        }
        console.log(chartData);

        console.log("#### Print Data TS file ####");
        const chartDataJson = JSON.stringify(chartData, null, 2);
        const articleListJson = JSON.stringify(pagingArticles, null, 2);
        const githubContributionsJson = JSON.stringify(
            gitHubContributions,
            null,
            2
        );
        const popularArticlesWithRankJson = JSON.stringify(
            popularArticlesWithRank,
            null,
            2
        );
        const noteArticlesJson = JSON.stringify(noteData, null, 2);
        const speakerDecksJson = JSON.stringify(decks, null, 2);
        const jsonContent = `export const TechArticleData = ${chartDataJson};\nexport const TechArticleList = ${articleListJson};\nexport const GitHubContributions = ${githubContributionsJson};\nexport const PopularArticles = ${popularArticlesWithRankJson};\nexport const noteArticles = ${noteArticlesJson};\nexport const speakerDecks = ${speakerDecksJson};`;
        fs.writeFile(FILE_PATH, jsonContent, "utf8", (err) => {
            if (err) {
                throw err;
            }
            console.log("#### Print Succeeded!! ####");
        });

        console.log("#### End make article data ####");
    } catch (e) {
        console.error("#### Error occured!!! ####");
        console.error("Error", e);
        return;
    }
};

const getZennArticles = async () => {
    const url =
        "https://zenn.dev/api/articles?username=yskn_sid25&order=latest&count=500";
    console.log("#### Zenn Data Fetching ####");
    return await fetch(url)
        .then((response) => {
            console.log("#### Zenn Data Responsed ####");
            return response.json();
        })
        .then((data) => {
            console.log("#### Zenn Data Fetched ####");
            return data;
        })
        .catch((error) => {
            console.error("Zenn Request failed", error);
            throw error;
        });
};

const getQiitaArticles = async (env) => {
    const url =
        "https://qiita.com/api/v2/items?per_page=100&query=user:ysknsid25";
    return await fetch(url, {
        headers: {
            Authorization: `Bearer ${env.parsed.QIITA_API_KEY}`,
        },
    })
        .then((response) => {
            return response.json();
        })
        .then((data) => {
            //console.log(data[0])
            return data;
        })
        .catch((error) => {
            console.error("Qiita Request failed", error);
            throw error;
        });
};

const getNoteArticles = async () => {
    let page = 1;
    let articles = [];
    while (true) {
        console.log(`fetch note page:${page}`);
        const url = `https://note.com/api/v2/creators/samurai_se/contents?kind=note&page=${page}`;
        const response = await fetch(url);
        const json = await response.json();
        const filteredData = json.data.contents.filter((content) => {
            const { year } = getDate(content.publishAt);
            const newDate = new Date();
            return !content.isPinned && year === newDate.getFullYear();
        });
        if (filteredData.length === 0) {
            break;
        }
        articles = articles.concat(filteredData);
        page++;
    }
    return articles.map((article) => {
        return {
            title: article.name,
            url: article.noteUrl,
        };
    });
};

const getSpeakerDecks = async () => {
    const url = "https://speakerdeck.com/ysknsid25.rss";
    // HTTPリクエストでRSSデータを取得
    const response = await fetch(url);
    const xmlData = await response.text();
    const parser = new xml2js.Parser();
    let returnData = [];
    parser.parseString(xmlData, (err, result) => {
        if (err) {
            console.error("XMLの解析に失敗しました:", err);
            return;
        }
        const filteredData = result.rss.channel[0].item
            .filter((data) => {
                const { year } = getDate(data.pubDate[0]);
                console.log(year);
                const today = new Date();
                return year === today.getFullYear();
            })
            .map((data) => {
                return {
                    title: data.title[0],
                    url: data.link[0],
                };
            });
        console.log(filteredData);
        returnData = filteredData;
    });
    return returnData;
};

const getGitHubControbutions = async () => {
    // 現在の日付を取得
    const currentDate = new Date();
    const before3date = new Date();
    // 3ヶ月前の日付を取得
    before3date.setMonth(before3date.getMonth() - 3);
    const currentDateISO = currentDate.toISOString().slice(0, 10);
    const before3MonthISO = before3date.toISOString().slice(0, 10);
    console.log(currentDateISO);
    console.log(before3MonthISO);

    const url = `https://github-contributions-api.deno.dev/ysknsid25.json?flat=true&from=${before3MonthISO}&to=${currentDateISO}`;
    return await fetch(url)
        .then((response) => {
            return response.json();
        })
        .then((data) => {
            return data;
        })
        .catch((error) => {
            console.error("GitHub Request failed", error);
            throw error;
        });
};

const getZennData = (articles) => {
    const zennData = {};
    articles.map((article) => {
        const key = getKey(article.published_at);
        if (zennData[key]) {
            zennData[key].articlesCount += 1;
            zennData[key].likeCount += article.liked_count;
        } else {
            zennData[key] = {
                articlesCount: 1,
                likeCount: article.liked_count,
            };
        }
    });
    return zennData;
};

const getQiitaData = (articles) => {
    const qiitaData = {};
    articles.map((article) => {
        const key = getKey(article.created_at);
        if (qiitaData[key]) {
            qiitaData[key].articlesCount += 1;
            qiitaData[key].likeCount += article.likes_count;
        } else {
            qiitaData[key] = {
                articlesCount: 1,
                likeCount: article.likes_count,
            };
        }
    });
    return qiitaData;
};

const getZennArticleList = (articles) => {
    return articles.map((article) => {
        const { year, month, day } = getDate(article.published_at);
        const formattedFullDate = getFormattedFullDate(year, month, day);
        return {
            treeType: "🖋",
            img: "zenn",
            year: formattedFullDate,
            title: article.title,
            url: `https://zenn.dev${article.path}`,
            content: `❤️ ${article.liked_count}`,
            likeCount: article.liked_count,
        };
    });
};

const getQiitaArticleList = (articles) => {
    return articles.map((article) => {
        const { year, month, day } = getDate(article.created_at);
        const formattedFullDate = getFormattedFullDate(year, month, day);
        return {
            treeType: "🖋",
            img: "qiita",
            year: formattedFullDate,
            title: article.title,
            url: article.url,
            content: `❤️ ${article.likes_count}`,
            likeCount: article.likes_count,
        };
    });
};

const getDate = (publishDateAt) => {
    const date = new Date(publishDateAt);
    const year = date.getFullYear();
    const month = date.getMonth() + 1;
    const day = date.getDate();
    return { year, month, day };
};

const getKey = (publishDateAt) => {
    const { year, month } = getDate(publishDateAt);
    return getFormattedDate(year, month);
};

const getFormattedDate = (year, month) => {
    return year + "/" + (month < 10 ? "0" : "") + month;
};

const getFormattedFullDate = (year, month, day) => {
    const formattedYearMonth = getFormattedDate(year, month);
    return formattedYearMonth + "/" + (day < 10 ? "0" : "") + day;
};

const comparePublishDate = (a, b) => {
    const dateA = new Date(a.year);
    const dateB = new Date(b.year);
    return dateB - dateA;
};

const compareLikeCount = (a, b) => {
    const likeA = new Date(a.year);
    const likeB = new Date(b.year);
    return likeB - likeA;
};

const getPagingArticles = (articleList) => {
    const pagingArticles = [];
    const perPage = 5;
    const pageCount = Math.ceil(articleList.length / perPage);
    for (let i = 0; i < pageCount; i++) {
        pagingArticles.push(articleList.slice(i * perPage, (i + 1) * perPage));
    }
    return pagingArticles;
};

await main();

File: make-tech-article-data.rb
require 'date'
require 'net/http'
require 'json'
require 'dotenv'
require 'fileutils'

BEGIN_YEAR = 2023
END_YEAR = 2024
BEGIN_MONTH = 4
FILE_PATH = './src/data/TechArticleData.ts'

def main
  puts '#### Begin make article data ####'

  puts '#### Loading Config... ####'
  Dotenv.load
  api_key = ENV['QIITA_API_KEY']
  puts "Qiita API Key: #{api_key}"

  begin
    puts '#### Fetch Zenn Info ####'
    original_zenn_data = get_zenn_articles
    zenn_data = get_zenn_data(original_zenn_data['articles'])
    zenn_article_list = get_zenn_article_list(original_zenn_data['articles'])
    puts zenn_article_list.length

    puts '#### Fetch Qiita Info ####'
    original_qiita_data = get_qiita_articles(api_key)
    qiita_data = get_qiita_data(original_qiita_data)
    qiita_article_list = get_qiita_article_list(original_qiita_data)
    puts qiita_article_list.length

    puts '#### Fetch GitHub Info ####'
    git_hub_data = get_git_hub_contributions
    git_hub_contributions = git_hub_data['contributions'].map do |contribution|
      {
        date: contribution['date'].gsub(/-/, '/'),
        contribution_count: contribution['contributionCount']
      }
    end
    puts git_hub_contributions.length

    # データをソートする
    puts '#### Make Sort And Paging Articles List ####'
    merged_and_sorted_article_list = (zenn_article_list + qiita_article_list).sort_by { |a| a[:year] }.reverse
    puts "Length is #{merged_and_sorted_article_list.length}"
    paging_articles = get_paging_articles(merged_and_sorted_article_list)
    puts "Page Size is #{paging_articles.length}"
    now_year = Time.now.year
    popular_articles = merged_and_sorted_article_list.select { |article| article[:year].include?(now_year.to_s) }
                                                     .sort_by { |a| -a[:like_count] }
    rank_emoji_array = ['🥇', '🥈', '🥉']
    sliced_popular_articles = popular_articles.length > 2 ? popular_articles.first(3) : popular_articles
    popular_articles_with_rank = sliced_popular_articles.each_with_index.map do |article, index|
      article.merge(treeType: rank_emoji_array[index])
    end

    puts '#### Make Chart Data ####'
    date = Time.now
    current_year = date.year
    current_month = date.month
    chart_data = {
      articlesCounts: [],
      yearFavoritesCounts: [],
      favoritesCounts: [],
      yearFavoritesCounts: []
    }

    (BEGIN_YEAR..END_YEAR).each do |year|
      year_article_count = { year: year.to_s, articles: 0 }
      year_favorites_count = { year: year.to_s, favorites: 0 }

      12.times do |month|
        next if year == BEGIN_YEAR && month + 1 < BEGIN_MONTH
        break if current_year == year && current_month < month + 1

        key = get_formatted_date(year, month + 1)
        articles_count = {
          yearMonth: key,
          zenn: zenn_data[key] ? zenn_data[key][:articles_count] : 0,
          qiita: qiita_data[key] ? qiita_data[key][:articles_count] : 0
        }
        chart_data[:articlesCounts] << articles_count
        year_article_count[:articles] += articles_count[:zenn] + articles_count[:qiita]

        favorites_count = {
          yearMonth: key,
          zenn: zenn_data[key] ? zenn_data[key][:like_count] : 0,
          qiita: qiita_data[key] ? qiita_data[key][:like_count] : 0
        }
        chart_data[:favoritesCounts] << favorites_count
        year_favorites_count[:favorites] += favorites_count[:zenn] + favorites_count[:qiita]
      end

      chart_data[:yearFavoritesCounts] << year_article_count
      chart_data[:yearFavoritesCounts] << year_favorites_count
    end

    puts chart_data

    puts '#### Print Data TS file ####'
    chart_data_json = JSON.pretty_generate(chart_data)
    article_list_json = JSON.pretty_generate(paging_articles)
    github_contributions_json = JSON.pretty_generate(git_hub_contributions)
    popular_articles_with_rank_json = JSON.pretty_generate(popular_articles_with_rank)

    json_content = <<~JSON
      export const TechArticleData = #{chart_data_json};
      export const TechArticleList = #{article_list_json};
      export const GitHubContributions = #{github_contributions_json};
      export const PopularArticles = #{popular_articles_with_rank_json};
    JSON

    File.write(FILE_PATH, json_content)
    puts '#### Print Succeeded!! ####'

    puts '#### End make article data ####'
  rescue => e
    puts '#### Error occurred!!! ####'
    puts "Error: #{e.message}"
  end
end

def get_zenn_articles
  url = URI('https://zenn.dev/api/articles?username=yskn_sid25&order=latest&count=500')
  response = Net::HTTP.get(url)
  JSON.parse(response)
rescue StandardError => e
  puts "Zenn Request failed: #{e.message}"
  raise e
end

def get_qiita_articles(api_key)
  url = URI('https://qiita.com/api/v2/items?per_page=100&query=user:ysknsid25')
  request = Net::HTTP::Get.new(url)
  request['Authorization'] = "Bearer #{api_key}"

  response = Net::HTTP.start(url.hostname, url.port, use_ssl: true) { |http| http.request(request) }
  JSON.parse(response.body)
rescue StandardError => e
  puts "Qiita Request failed: #{e.message}"
  raise e
end

def get_git_hub_contributions
  current_date = Date.today
  before_3_months_date = current_date << 3
  current_date_iso = current_date.strftime('%Y-%m-%d')
  before_3_months_iso = before_3_months_date.strftime('%Y-%m-%d')

  url = URI("https://github-contributions-api.deno.dev/ysknsid25.json?flat=true&from=#{before_3_months_iso}&to=#{current_date_iso}")
  response = Net::HTTP.get(url)
  JSON.parse(response)
rescue StandardError => e
  puts "GitHub Request failed: #{e.message}"
  raise e
end

def get_zenn_data(articles)
  zenn_data = {}
  articles.each do |article|
    key = get_key(article['published_at'])
    if zenn_data[key]
      zenn_data[key][:articles_count] += 1
      zenn_data[key][:like_count] += article['liked_count']
    else
      zenn_data[key] = { articles_count: 1, like_count: article['liked_count'] }
    end
  end
  zenn_data
end

def get_qiita_data(articles)
  qiita_data = {}
  articles.each do |article|
    key = get_key(article['created_at'])
    if qiita_data[key]
      qiita_data[key][:articles_count] += 1
      qiita_data[key][:like_count] += article['likes_count']
    else
      qiita_data[key] = { articles_count: 1, like_count: article['likes_count'] }
    end
  end
  qiita_data
end

def get_zenn_article_list(articles)
  articles.map do |article|
    year, month, day = get_date(article['published_at'])
    formatted_full_date = get_formatted_full_date(year, month, day)
    {
      treeType: '🖋',
      img: 'zenn',
      year: formatted_full_date,
      title: article['title'],
      url: "https://zenn.dev#{article['path']}",
      content: "❤️ #{article['liked_count']}",
      like_count: article['liked_count']
    }
  end
end

def get_qiita_article_list(articles)
  articles.map do |article|
    year, month, day = get_date(article['created_at'])
    formatted_full_date = get_formatted_full_date(year, month, day)
    {
      treeType: '🖋',
      img: 'qiita',
      year: formatted_full_date,
      title: article['title'],
      url: article['url'],
      content: "❤️ #{article['likes_count']}",
      like_count: article['likes_count']
    }
  end
end

def get_date(publish_date_at)
  date = DateTime.parse(publish_date_at)
  year = date.year
  month = date.month
  day = date.day
  [year, month, day]
end

def get_key(publish_date_at)
  year, month, = get_date(publish_date_at)
  get_formatted_date(year, month)
end

def get_formatted_date(year, month)
  "#{year}/#{format('%02d', month)}"
end

def get_formatted_full_date(year, month, day)
  formatted_year_month = get_formatted_date(year, month)
  "#{formatted_year_month}/#{format('%02d', day)}"
end

def get_paging_articles(article_list)
  paging_articles = []
  per_page = 5
  page_count = (article_list.length / per_page.to_f).ceil
  page_count.times do |i|
    paging_articles << article_list.slice(i * per_page, per_page)
  end
  paging_articles
end

main
